### ISSUE TITLE
Conflicting definitions for Set constructor causes unexpected default generic in TS 3.5

### ISSUE THREAD

**Issue by evmar** [microsoft/TypeScript#31990]
**TypeScript Version:**  3.4.5 vs versions after 3.5

**Search Terms:** set constructor setconstructor unknown

**Code**

```ts
let s = new Set();
```

**Expected behavior:**

In TS 3.4: we expect the default chosen param to give us a `Set<{}>`.
In TS 3.5: we expect a `Set<unknown>`.
That change was an intended change in TS 3.5.

**Actual behavior:**

In TS 3.4, this actually was a `Set<any>`!

It appears there are multiple overloads of the Set constructor.

lib.es2015.collection.d.ts says:

```
    new <T = any>(values?: ReadonlyArray<T> | null): Set<T>;
```

while lib.es2105.iterable.d.ts says:

```
    new <T>(iterable?: Iterable<T> | null): Set<T>;
```

Note that one has `T = any` and the other doesn't.

So somehow TS 3.4 vs TS 3.5 decided to change whether to obey the `any` default, I think?

**Comment by evmar** [comment]
@weswigham I think this might be unintended fallout of your default type parameter constraints change, maybe?  https://github.com/microsoft/TypeScript/pull/30637

I think the two constructors should at least be consistent, but I'm not sure which is preferred, probably the `any`-less one?

**Comment by weswigham** [comment]
> I'm not sure which is preferred, probably the any-less one?

Yeah, I'm not sure why we even have aversion that defaults to `any` - the's obviously quite unsafe.

**Comment by fatcerberus** [comment]
`Array` has that problem too:

```ts
let arr = new Array(10);  // any[]
```

**Comment by weswigham** [comment]
@DanielRosenwasser do we want to change our behavior here? I think we last looked at it some time around https://github.com/microsoft/TypeScript/pull/25374

**Comment by evmar** [comment]
Perhaps related, I notice in 3.4->3.5, SetConstructor changed from

```
   new <T>(iterable: Iterable<T>): Set<T>;
```

to

```
    new <T>(iterable?: Iterable<T> | null): Set<T>;
```

So perhaps that is changing which overload is selected when.

**Comment by DanielRosenwasser** [comment]
Yeah, it's unsafe, but it would be incredibly annoying to break people.

This sort of comes back to the "I want this to be an 'implicit any' for the `noImplicitAny` scenarios."

**Comment by fatcerberus** [comment]
I guess it can’t really be `unknown[]` since then it wouldn’t be assignable to anything but it also can’t reasonably be `never[]` because then you can’t put stuff in it (unless it could be made to work like evolving array types).

**Comment by evmar** [comment]
Just to restate, what I am concerned about here is whether "new Set()" is supposed to give you a "Set<any>" or a "Set<unknown>", given the conflicting definitions in the typings.  I believe the intention was for 3.5 to make it <unknown> and for that to break apps, but I think the resolution of this bug is to make the typings express that intent more clearly.

**Comment by evmar** [comment]
Circling back on this -- I think maybe I failed to communicate the bug here.

In practice, TS3.4 => TS3.5 changed what `new Set()` gives you, from `Set<any>` to `Set<unknown>`.  (That is too late to undo, though massively annoying for us to upgrade through.)

But the actual bug here is that I think you didn't change the definition of Set!  It still says `any` in the typings, something else must be going wrong.
